### ISSUE TITLE
Allow ratcheting validation on custom resources

### ISSUE THREAD

**Issue by msau42** [kubernetes/kubernetes#94060]
(hoisted from https://github.com/kubernetes/kubernetes/issues/94060#issuecomment-677644185)

To support improving validation on a custom resource definition when there is existing persisted custom resource data that only passes the old less restrictive validation, allow a CR update that fails custom CRD schema validation IFF the existing object fails the custom CRD schema validation with identical errors (object-level ratcheting)

---

original description:

**What would you like to be added**:
The ability to make updates to certain fields that are used in the deletion flow (like (ownerReferences, finalizers, managedFields) ), on objects that would otherwise fail validation.

**Why is this needed**:
For the VolumeSnapshot CRDs, we wanted to introduce stricter validation on some fields, but found it very difficult to accomplish this in a way that could retain backwards compatibility. The procedure we came up with was:

1. Introduce a validating webhook for an existing v1beta1 version that does ratcheting validation - existing invalid objects continue to be invalid while new objects or updates to valid objects that would cause them to be invalid are rejected. Proactively detect and reach out to consumers to clean up their invalid objects.
2. 1+ releases later, add v1 as a serving version, but keep v1beta as the storage version, so that we have the ability to rollback the CRD schema to drop v1 if that's causing trouble in the controllers.
3. Once we are 100% sure no invalid objects exist in the cluster, update the storage version to v1.

The main challenge right now is that we cannot upgrade the storage version to v1 until we are absolutely sure no invalid objects exist in the cluster. Otherwise, updates like removing finalizers would fail validation.

xref: https://github.com/kubernetes/enhancements/pull/1902#discussion_r468782652

@kubernetes/sig-api-machinery-feature-requests @kubernetes/sig-storage-misc

**Comment by lavalamp** [comment]
I'm a little confused, objects have to be round-trippable, so why can't you continue to keep the same validation for v1 objects?

**Comment by AndiLi99** [comment]
> I'm a little confused, objects have to be round-trippable, so why can't you continue to keep the same validation for v1 objects?

The goal is to change the validation for v1 objects, to make it more strict.

**Comment by lavalamp** [comment]
Validation is somewhat version independent. Introducing a new version doesn't actually help this problem, and possibly makes it harder. All objects must be representable in all versions.

E.g., the OP says "we cannot upgrade the storage version to v1 until we are absolutely sure no invalid objects"

But this isn't the case. You can't add v1 if it cannot represent the invalid objects. Because, for example, we have the property that clients can write an "invalid" object in v1beta1 and then immediately read it from v1. So, if you add v1, and it can represent the invalid objects, then it can also store those objects.

The request that e.g. metadata writes go through even on invalid objects that aren't otherwise changed seems worth considering. But I'm not following the reasoning given.

**Comment by liggitt** [comment]
I don't think the storage version is really relevant. The primary goal is to be able to introduce a new API version with improved schema validation.

This is the problematic scenario:
1. have a v1beta1 version with insufficient validation (extremely common)
2. use the v1beta1 API to persist insufficiently validated instances (extremely common)
3. realize the error of your ways, and add a v1 version with good schema validation

Outcome:
* v1beta1 clients can continue to use the v1beta1 API without being broken
* v1 clients doing a read/write of the invalid instances will get rejected with schema validation errors, even when doing a no-op write (or only changing metadata fields). This breaks generic controllers which use the preferred version of a resource and will use v1 over v1beta1 (storage migrator, garbage collector controller, etc).

Possible solutions:
1. allow a CR update that fails custom CRD schema validation IFF it only changes finalizers/ownerReferences/managedFields (no ratcheting, just targeting unblocking deletion/GC flows)
2. allow a CR update that fails custom CRD schema validation IFF it only changes finalizers/ownerReferences/managedFields AND the existing object also fails the custom CRD schema validation (ratcheting, targeting unblocking deletion/GC flows)
3. allow a CR update that fails custom CRD schema validation iff the existing object fails the custom CRD schema validation with identical errors (object-level ratcheting)

In the abstract, I prefer option 3.

**Comment by liggitt** [comment]
my observation is that the current implications / failure modes of adding a new stricter version:
* bites people who don't understand the implications who are trying to do the right thing and add an improved version
* discourages people who do understand the implications from adding new versions with improved validation

**Comment by apelisse** [comment]
I agree with Jordan, the current incentives are wrong. People should be able to improve their schemas.

**Comment by apelisse** [comment]
Going a little off-topic here, also note that this doesn't solve the problem of objects "template" that create other objects (deployment -> pods). We still can't change the validation on the podtemplate.

**Comment by liggitt** [comment]
> Going a little off-topic here, also note that this doesn't solve the problem of objects "template" that create other objects (deployment -> pods). We still can't change the validation on the podtemplate.

one problem at a time :-)

**Comment by apelisse** [comment]
I don't disagree, but it might be a good time to look at the problem holistically :-)

**Comment by sttts** [comment]
Have we thought about start having ratcheting validation in CRDs?

```
type: string
x-kubernetes-ratcheting-schema:
  minLength: 1
```

MinLength would be enforced only when it was true already before.

**Comment by liggitt** [comment]
> Have we thought about start having ratcheting validation in CRDs?

yes. from https://github.com/kubernetes/kubernetes/issues/94060#issuecomment-677644185:

> Possible solutions:
> 
> ...
> 3. allow a CR update that fails custom CRD schema validation IFF the existing object fails the custom CRD schema validation with identical errors (object-level ratcheting)
>
> In the abstract, I prefer option 3.

**Comment by liggitt** [comment]
This is important to allow custom resource authors to safely improve schemas that were previously poorly defined

**Comment by liggitt** [comment]
yup, closing in favor of that

/close

**Comment by caozhuozi** [comment]
> v1 clients doing a read/write of the invalid instances will get rejected with schema validation errors, even when doing a no-op write (or only changing metadata fields).

is this because, lets say read, after conversion v1beta1 -> v1, the tighter validation of v1 will reject the converted v1?
